src/optim.jl

```diff
@@ -317,3 +322,127 @@ function lower_ptls!(mod::LLVM.Module)
+            isdeclaration(candidate_f) && continue
+            for use in uses(candidate_f)
+                inst = user(use)
+                bb = LLVM.parent(inst)
+                f = LLVM.parent(bb)
+                if f in worklist
+                    push!(worklist, candidate_f)
+                end
+            end
+        end
+    end
+
+    # intrinsic returning an opaque pointer to the kernel state.
+    # this is both for extern uses, and to make this transformation a two-step process.
+    state_typ = LLVM.PointerType(LLVM.StructType(LLVMType[]; ctx))
+    state_getter = if haskey(functions(mod), "julia.gpu.state_getter")
+        functions(mod)["julia.gpu.state_getter"]
+    else
+        LLVM.Function(mod, "julia.gpu.state_getter", LLVM.FunctionType(state_typ))
+    end
+    push!(function_attributes(state_getter), EnumAttribute("readnone", 0; ctx))
+
+    # add a state argument to every function
+    replaced = Set{LLVM.Function}()
+    for f in worklist
+        fn = LLVM.name(f)
+        ft = eltype(llvmtype(f))
+
+        # create a new function
+        new_param_types = [state_typ, parameters(ft)...]
+        new_ft = LLVM.FunctionType(return_type(ft), new_param_types)
+        new_f = LLVM.Function(mod, "", new_ft)
+        LLVM.name!(parameters(new_f)[1], "state")
+        linkage!(new_f, linkage(f))
+
+        # clone
+        value_map = Dict{LLVM.Value, LLVM.Value}()
+        for (param, new_param) in zip(parameters(f), parameters(new_f)[2:end])
+            LLVM.name!(new_param, LLVM.name(param))
+            value_map[param] = new_param
+        end
+        clone_into!(new_f, f; value_map,
+                    changes=LLVM.API.LLVMCloneFunctionChangeTypeGlobalChanges)
+        # NOTE: we need global changes because LLVM 12 wants to clone debug metadata
+
+        # update uses
+        Builder(ctx) do builder
+            for use in uses(f)
+                # only replace calls
+                # TODO: can we have other uses?
+                inst = user(use)
+                inst isa LLVM.CallInst || inst isa LLVM.InvokeInst || inst isa LLVM.CallBrInst || continue
+
+                # NOTE: we unconditionally add the state argument, even if there's no uses,
+                #       assuming we'll perform dead arg elimination during optimization.
+
+                # forward the state argument
+                position!(builder, inst)
+                state = call!(builder, state_getter, Value[], "state")
+                new_inst = call!(builder, new_f, [state, operands(inst)[1:end-1]...])
+
+                replace_uses!(inst, new_inst)
+                @assert isempty(uses(inst))
+                unsafe_delete!(LLVM.parent(inst), inst)
+            end
+        end
+
+        # clean-up
+        if f == entry
+            entry = new_f
+        end
+        @assert isempty(uses(f))
+        unsafe_delete!(mod, f)
+        LLVM.name!(new_f, fn)
+        push!(replaced, new_f)
+    end
+    empty!(worklist)
```

Comment:
Won't the GC just delete `worklist` for us once it goes out of scope?

Reply:
Yeah, I added this after wrongfully using entries from the array (which are `unsafe_delete!`d so invalid) :smile: